In [296]:
import pandas as pd
import numpy as np


# EDA

### Reading the data

In [297]:
data = pd.read_csv('data/train.csv')
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  ...  1stFlrSF  \
0          GLQ         706          Unf           0        150  ...       856   
1          ALQ         978          Unf           0        284  ...      1262   
2          GLQ         486          Unf           0        434  ...       920   
3          ALQ         216          Unf           0        540  ...       961   
4          GLQ         655          Unf           0        490  ...      1145   

  2ndFlrSF LowQualFinSF GrLivArea BsmtFullBath  BsmtHalfBath  FullBath  \
0      854            0      1710            1             0         2   
1        0            0      1262            0             1         2   
2      866            0      1786            1             0         2   
3      756            0      1717            1             0         1   
4     1053            0      2198            1             0         2   

   HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  \
0         1             3             1           Gd             8   
1         0             3             1           TA             6   
2         1             3             1           Gd             6   
3         0             3             1           Gd             7   
4         1             4             1           Gd             9   

   Functional  Fireplaces  FireplaceQu GarageType  GarageYrBlt GarageFinish  \
0         Typ           0          NaN     Attchd       2003.0          RFn   
1         Typ           1           TA     Attchd       1976.0          RFn   
2         Typ           1           TA     Attchd       2001.0          RFn   
3         Typ           1           Gd     Detchd       1998.0          Unf   
4         Typ           1     

In [298]:
data.columns # checking the features.

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [299]:
data.drop('Id', axis=1, inplace=True) # removing Id feature (will not give any info on the price just fifo).
data

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0             60       RL         65.0     8450   Pave   NaN      Reg   
1             20       RL         80.0     9600   Pave   NaN      Reg   
2             60       RL         68.0    11250   Pave   NaN      IR1   
3             70       RL         60.0     9550   Pave   NaN      IR1   
4             60       RL         84.0    14260   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
1455          60       RL         62.0     7917   Pave   NaN      Reg   
1456          20       RL         85.0    13175   Pave   NaN      Reg   
1457          70       RL         66.0     9042   Pave   NaN      Reg   
1458          20       RL         68.0     9717   Pave   NaN      Reg   
1459          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
1455         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1456         Lvl    AllPub    Inside       Gtl       NWAmes       Norm   
1457         Lvl    AllPub    Inside       Gtl      Crawfor       Norm   
1458         Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1459         Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
1455       Norm     1Fam     2Story            6            5       1999   
1456       Norm     1Fam     1Story            6            6       1978   
1457       Norm     1Fam     2Story            7            9       1941   
1458       Norm     1Fam     1Story            5            6       1950   
1459       Norm     1Fam     1Story            5            6       1965   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd       None   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
1455          2000     Gable  CompShg     VinylSd     VinylSd       None   
1456          1988     Gable  CompShg     Plywood     Plywood      Stone   
1457          2006     Gable  CompShg     CemntBd     CmentBd       None   
1458          1996       Hip  CompShg     MetalSd     MetalSd       None   
1459          1965     Gable  CompShg     HdBoard     HdBoard       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   
4          350.0        Gd        TA      PConc       Gd       TA   


### Check for nan values and explore.

we can see:

- PoolQC         0.995205
- MiscFeature    0.963014
- Alley          0.937671
- Fence          0.807534

we will drop those features at first.

In [300]:
data.isna().mean().sort_values(ascending=False)

PoolQC         0.995205
MiscFeature    0.963014
Alley          0.937671
Fence          0.807534
FireplaceQu    0.472603
                 ...   
Heating        0.000000
HeatingQC      0.000000
MSZoning       0.000000
1stFlrSF       0.000000
SalePrice      0.000000
Length: 80, dtype: float64

##### Fixing na values in PoolQC column 

In [301]:
# data['PoolQC'] = data['PoolQC'].fillna('NA')
# len(data[(data['PoolArea'] == 0) & (data['PoolQC'] !='NA')])
# data.head(5)

In [302]:
data.drop(['PoolQC', 'MiscFeature', 'Alley', 'Fence'], axis=1, inplace=True)
data.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          60       RL         65.0     8450   Pave      Reg         Lvl   
1          20       RL         80.0     9600   Pave      Reg         Lvl   
2          60       RL         68.0    11250   Pave      IR1         Lvl   
3          70       RL         60.0     9550   Pave      IR1         Lvl   
4          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   

  RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond  \
0  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA   
1  CompShg     MetalSd     MetalSd       None         0.0        TA        TA   
2  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA   
3  CompShg     Wd Sdng     Wd Shng       None         0.0        TA        TA   
4  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd        TA   

  Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1  \
0      PConc       Gd       TA           No          GLQ         706   
1     CBlock       Gd       TA           Gd          ALQ         978   
2      PConc       Gd       TA           Mn          GLQ         486   
3     BrkTil       TA       Gd           No          ALQ         216   
4      PConc       Gd       TA           Av          GLQ         655   

  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC  \
0          Unf           0        150          856    GasA        Ex   
1          Unf           0        284         1262    GasA        Ex   
2          Unf           0        434          920    GasA        Ex   
3          Unf           0        540          756    GasA        Gd   
4          Unf           0        490         1145    GasA        Ex   

  CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  \
0          Y      SBrkr       856       854             0       1710   
1          Y      SBrkr      1262         0             0       1262   
2          Y      SBrkr       920       866             0       1786   
3          Y      SBrkr       961       756             0       1717   
4          Y      SBrkr      1145      1053             0       2198   

   BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0             1             0         2         1             3             1   
1             0             1         2         0             3             1   
2             1             0         2         1             3             1   
3             1             0         1         0             3             1   
4             1             0         2         1             4             1   

  KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  \
0          Gd             8        Typ           0         NaN     Attchd   
1          TA             6        Typ           1          TA     Attchd   
2          Gd             6        Typ           1          TA     Attchd   
3          Gd             7

In [303]:
print(data.isna().mean().sort_values(ascending=False).head(20) )

FireplaceQu     0.472603
LotFrontage     0.177397
GarageType      0.055479
GarageYrBlt     0.055479
GarageFinish    0.055479
GarageQual      0.055479
GarageCond      0.055479
BsmtExposure    0.026027
BsmtFinType2    0.026027
BsmtQual        0.025342
BsmtCond        0.025342
BsmtFinType1    0.025342
MasVnrType      0.005479
MasVnrArea      0.005479
Electrical      0.000685
Fireplaces      0.000000
KitchenQual     0.000000
KitchenAbvGr    0.000000
BedroomAbvGr    0.000000
HalfBath        0.000000
dtype: float64


In [304]:
print(data.isna().sum().sort_values(ascending=False).head(20))

FireplaceQu     690
LotFrontage     259
GarageType       81
GarageYrBlt      81
GarageFinish     81
GarageQual       81
GarageCond       81
BsmtExposure     38
BsmtFinType2     38
BsmtQual         37
BsmtCond         37
BsmtFinType1     37
MasVnrType        8
MasVnrArea        8
Electrical        1
Fireplaces        0
KitchenQual       0
KitchenAbvGr      0
BedroomAbvGr      0
HalfBath          0
dtype: int64


In [305]:
data['FireplaceQu'] = data['FireplaceQu'].fillna("NA") # replacing nans with new category "NA"

##### Fixing na values in LotFrontage column - disc: Linear feet of street connected to property

there are 1201 rows with null values, are missing 259.
 - the question is if null value is equal to 0 ?? 

In [306]:
data[data['LotFrontage'].isna()]

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
7             60       RL          NaN    10382   Pave      IR1         Lvl   
12            20       RL          NaN    12968   Pave      IR2         Lvl   
14            20       RL          NaN    10920   Pave      IR1         Lvl   
16            20       RL          NaN    11241   Pave      IR1         Lvl   
24            20       RL          NaN     8246   Pave      IR1         Lvl   
...          ...      ...          ...      ...    ...      ...         ...   
1429          20       RL          NaN    12546   Pave      IR1         Lvl   
1431         120       RL          NaN     4928   Pave      IR1         Lvl   
1441         120       RM          NaN     4426   Pave      Reg         Lvl   
1443          30       RL          NaN     8854   Pave      Reg         Lvl   
1446          20       RL          NaN    26142   Pave      IR1         Lvl   

     Utilities LotConfig LandSlope Neighborhood Condition1 Condition2  \
7       AllPub    Corner       Gtl       NWAmes       PosN       Norm   
12      AllPub    Inside       Gtl       Sawyer       Norm       Norm   
14      AllPub    Corner       Gtl        NAmes       Norm       Norm   
16      AllPub   CulDSac       Gtl        NAmes       Norm       Norm   
24      AllPub    Inside       Gtl       Sawyer       Norm       Norm   
...        ...       ...       ...          ...        ...        ...   
1429    AllPub    Corner       Gtl       NWAmes       Norm       Norm   
1431    AllPub    Inside       Gtl      NPkVill       Norm       Norm   
1441    AllPub    Inside       Gtl      CollgCr       Norm       Norm   
1443    AllPub    Inside       Gtl      BrkSide       Norm       Norm   
1446    AllPub   CulDSac       Gtl      Mitchel       Norm       Norm   

     BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
7        1Fam     2Story            7            6       1973          1973   
12       1Fam     1Story            5            6       1962          1962   
14       1Fam     1Story            6            5       1960          1960   
16       1Fam     1Story            6            7       1970          1970   
24       1Fam     1Story            5            8       1968          2001   
...       ...        ...          ...          ...        ...           ...   
1429     1Fam     1Story            6            7       1981          1981   
1431   TwnhsE     1Story            6            6       1976          1976   
1441   TwnhsE     1Story            6            5       2004          2004   
1443     1Fam     1.5Unf            6            6       1916          1950   
1446     1Fam     1Story            5            7       1962          1962   

     RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea  \
7        Gable  CompShg     HdBoard     HdBoard      Stone       240.0   
12         Hip  CompShg     HdBoard     Plywood       None         0.0   
14         Hip  CompShg     MetalSd     MetalSd    BrkFace       212.0   
16       Gable  CompShg     Wd Sdng     Wd Sdng    BrkFace       180.0   
24       Gable  CompShg     Plywood     Plywood       None         0.0   
...        ...      ...         ...         ...        ...         ...   
1429     Gable  CompShg     MetalSd     MetalSd    BrkFace       310.0   
1431     Gable  CompShg     Plywood     Plywood       None         0.0   
1441     Gable  CompShg     VinylSd     VinylSd    BrkFace       147.0   
1443     Gable  CompShg     Wd Sdng     Wd Sdng       None         0.0   
1446     Gable  CompShg     HdBoard     HdBoard    BrkFace       189.0   

     ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
7           TA        TA     CBlock       Gd       TA           Mn   
12          TA        TA     CBlock       TA       TA           No   
14          TA        TA     CBlock       TA       TA           No   
16          TA        TA     CBlock       TA       TA         

In [307]:
len(data.index[data['GarageType'].isna()   &
               data['GarageYrBlt'].isna()  & 
               data['GarageFinish'].isna() &
               data['GarageQual'].isna()   &
               data['GarageCond'].isna()])

81

In [308]:
pd.set_option('display.max_columns', 76)
df_garage_null=data[data['GarageType'].isna()   &
               data['GarageYrBlt'].isna()  & 
               data['GarageFinish'].isna() &
               data['GarageQual'].isna()   &
               data['GarageCond'].isna()]
print("correlation GarageArea & df_garage_null" , len(df_garage_null[df_garage_null['GarageArea'] ==0]))
df_garage_null

correlation GarageArea & df_garage_null 81


MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
39            90       RL         65.0     6040   Pave      Reg         Lvl   
48           190       RM         33.0     4456   Pave      Reg         Lvl   
78            90       RL         72.0    10778   Pave      Reg         Lvl   
88            50  C (all)        105.0     8470   Pave      IR1         Lvl   
89            20       RL         60.0     8070   Pave      Reg         Lvl   
...          ...      ...          ...      ...    ...      ...         ...   
1349          70       RM         50.0     5250   Pave      Reg         Lvl   
1407          20       RL          NaN     8780   Pave      IR1         Lvl   
1449         180       RM         21.0     1533   Pave      Reg         Lvl   
1450          90       RL         60.0     9000   Pave      Reg         Lvl   
1453          20       RL         90.0    17217   Pave      Reg         Lvl   

     Utilities LotConfig LandSlope Neighborhood Condition1 Condition2  \
39      AllPub    Inside       Gtl      Edwards       Norm       Norm   
48      AllPub    Inside       Gtl      OldTown       Norm       Norm   
78      AllPub    Inside       Gtl       Sawyer       Norm       Norm   
88      AllPub    Corner       Gtl       IDOTRR      Feedr      Feedr   
89      AllPub    Inside       Gtl      CollgCr       Norm       Norm   
...        ...       ...       ...          ...        ...        ...   
1349    AllPub    Inside       Gtl      OldTown       Norm       Norm   
1407    AllPub    Corner       Gtl      Mitchel       Norm       Norm   
1449    AllPub    Inside       Gtl      MeadowV       Norm       Norm   
1450    AllPub       FR2       Gtl        NAmes       Norm       Norm   
1453    AllPub    Inside       Gtl      Mitchel       Norm       Norm   

     BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
39     Duplex     1Story            4            5       1955          1955   
48     2fmCon     2Story            4            5       1920          2008   
78     Duplex     1Story            4            5       1968          1968   
88       1Fam     1.5Fin            3            2       1915          1982   
89       1Fam     1Story            4            5       1994          1995   
...       ...        ...          ...          ...        ...           ...   
1349     1Fam     2Story            8            5       1872          1987   
1407     1Fam     1Story            5            5       1985          1985   
1449    Twnhs     SFoyer            5            7       1970          1970   
1450   Duplex     2Story            5            5       1974          1974   
1453     1Fam     1Story            5            5       2006          2006   

     RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea  \
39       Gable  CompShg     AsbShng     Plywood       None         0.0   
48       Gable  CompShg     MetalSd     MetalSd       None         0.0   
78         Hip  CompShg     HdBoard     HdBoard       None         0.0   
88         Hip  CompShg     Plywood     Plywood       None         0.0   
89       Gable  CompShg     VinylSd     VinylSd       None         0.0   
...        ...      ...         ...         ...        ...         ...   
1349     Gable  CompShg     MetalSd     MetalSd       None         0.0   
1407     Gable  CompShg     HdBoard     Plywood       None         0.0   
1449     Gable  CompShg     CemntBd     CmentBd       None         0.0   
1450     Gable  CompShg     VinylSd     VinylSd       None         0.0   
1453     Gable  CompShg     VinylSd     VinylSd       None         0.0   

     ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
39          TA        TA      PConc      NaN      NaN          NaN   
48          TA        TA     BrkTil       TA       TA           No   
78          TA        TA     CBlock       TA       TA           No   
88          Fa        Fa     CBlock       TA       Fa         

as we explored those null values are because there is no garage to the property.
we assign them as NA -> not availble. 
* categorical: 
    - GarageType
    - GarageQual	
    - GarageCond
    - GarageFinish
* numerical:
    - GarageYrBlt

In [309]:
data.fillna({'GarageType':'NA', 'GarageFinish':'NA', 'GarageQual':'NA', 'GarageCond':'NA', 'GarageYrBlt':.0}, inplace=True)
data.head(10)

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          60       RL         65.0     8450   Pave      Reg         Lvl   
1          20       RL         80.0     9600   Pave      Reg         Lvl   
2          60       RL         68.0    11250   Pave      IR1         Lvl   
3          70       RL         60.0     9550   Pave      IR1         Lvl   
4          60       RL         84.0    14260   Pave      IR1         Lvl   
5          50       RL         85.0    14115   Pave      IR1         Lvl   
6          20       RL         75.0    10084   Pave      Reg         Lvl   
7          60       RL          NaN    10382   Pave      IR1         Lvl   
8          50       RM         51.0     6120   Pave      Reg         Lvl   
9         190       RL         50.0     7420   Pave      Reg         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   
5    AllPub    Inside       Gtl      Mitchel       Norm       Norm     1Fam   
6    AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam   
7    AllPub    Corner       Gtl       NWAmes       PosN       Norm     1Fam   
8    AllPub    Inside       Gtl      OldTown     Artery       Norm     1Fam   
9    AllPub    Corner       Gtl      BrkSide     Artery     Artery   2fmCon   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   
5     1.5Fin            5            5       1993          1995     Gable   
6     1Story            8            5       2004          2005     Gable   
7     2Story            7            6       1973          1973     Gable   
8     1.5Fin            7            5       1931          1950     Gable   
9     1.5Unf            5            6       1939          1950     Gable   

  RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond  \
0  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA   
1  CompShg     MetalSd     MetalSd       None         0.0        TA        TA   
2  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA   
3  CompShg     Wd Sdng     Wd Shng       None         0.0        TA        TA   
4  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd        TA   
5  CompShg     VinylSd     VinylSd       None         0.0        TA        TA   
6  CompShg     VinylSd     VinylSd      Stone       186.0        Gd        TA   
7  CompShg     HdBoard     HdBoard      Stone       240.0        TA        TA   
8  CompShg     BrkFace     Wd Shng       None         0.0        TA        TA   
9  CompShg     MetalSd     MetalSd       None         0.0        TA        TA   

  Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1  \
0      PConc       Gd       TA           No          GLQ         706   
1     CBlock       Gd       TA           Gd          ALQ         978   
2      PConc       Gd       TA           Mn          GLQ         486   
3     BrkTil       TA       Gd           No          ALQ         216   
4      PConc       Gd       TA           Av          GLQ         655   
5       Wood       Gd       TA           No          GLQ         732   
6      PConc       Ex       TA           Av         

BsmtExposure     38
BsmtFinType2     38
BsmtQual         37
BsmtCond         37
BsmtFinType1     37

In [310]:
data[
    data['BsmtExposure'].isna() &
    ~data['BsmtFinType2'].isna() &
    ~data['BsmtQual'].isna() &
    ~data['BsmtCond'].isna() &
    ~data['BsmtFinType1'].isna()].index 

data.loc[[948]]

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
948          60       RL         65.0    14006   Pave      IR1         Lvl   

    Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
948    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   

    HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
948     2Story            7            5       2002          2002     Gable   

    RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual  \
948  CompShg     VinylSd     VinylSd    BrkFace       144.0        Gd   

    ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  \
948        TA      PConc       Gd       TA          NaN          Unf   

     BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating  \
948           0          Unf           0        936          936    GasA   

    HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
948        Ex          Y      SBrkr       936       840             0   

     GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
948       1776             0             0         2         1             3   

     KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces  \
948             1          Gd             7        Typ           1   

    FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea  \
948          TA     Attchd       2002.0          RFn           2         474   

    GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  \
948         TA         TA          Y         144           96              0   

     3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold SaleType  \
948          0            0         0        0       2    2006       WD   

    SaleCondition  SalePrice  
948        Normal     192500

line 948 can be predicted as a label using all other non-null rows of features: 
['BsmtFinType2']
['BsmtQual']
['BsmtCond']
['BsmtFinType1']

In [311]:
# to fill : data.loc[[948]]
df_bsmt_nn = data[
    ~data['BsmtExposure'].isna() &
    ~data['BsmtFinType2'].isna() &
    ~data['BsmtQual'].isna() &
    ~data['BsmtCond'].isna() &
    ~data['BsmtFinType1'].isna()]


In [315]:
exp_data = df_bsmt_nn[['BsmtFinType2', 'BsmtFinType2', 'BsmtQual', 'BsmtCond', 'BsmtFinType1', 'BsmtExposure']]
exp_labels = df_bsmt_nn['BsmtExposure']

In [316]:
from sklearn.preprocessing import OneHotEncoder

# pd.get_dummies(df_bsmt_nn['BsmtExposure'])
exp_data = pd.get_dummies(exp_data, columns = ['BsmtFinType2', 'BsmtFinType2', 'BsmtQual', 'BsmtCond', 'BsmtFinType1'], drop_first=True)
exp_labels =  pd.get_dummies(exp_labels, columns = ['BsmtExposure'])
exp_data

BsmtExposure  BsmtFinType2_BLQ  BsmtFinType2_GLQ  BsmtFinType2_LwQ  \
0              No                 0                 0                 0   
1              Gd                 0                 0                 0   
2              Mn                 0                 0                 0   
3              No                 0                 0                 0   
4              Av                 0                 0                 0   
...           ...               ...               ...               ...   
1455           No                 0                 0                 0   
1456           No                 0                 0                 0   
1457           No                 0                 0                 0   
1458           Mn                 0                 0                 0   
1459           No                 0                 0                 1   

      BsmtFinType2_Rec  BsmtFinType2_Unf  BsmtFinType2_BLQ  BsmtFinType2_GLQ  \
0                    0                 1                 0                 0   
1                    0                 1                 0                 0   
2                    0                 1                 0                 0   
3                    0                 1                 0                 0   
4                    0                 1                 0                 0   
...                ...               ...               ...               ...   
1455                 0                 1                 0                 0   
1456                 1                 0                 0                 0   
1457                 0                 1                 0                 0   
1458                 1                 0                 0                 0   
1459                 0                 0                 0                 0   

      BsmtFinType2_LwQ  BsmtFinType2_Rec  BsmtFinType2_Unf  BsmtFinType2_BLQ  \
0                    0                 0                 1                 0   
1                    0                 0                 1                 0   
2                    0                 0                 1                 0   
3                    0                 0                 1                 0   
4                    0                 0                 1                 0   
...                ...               ...               ...               ...   
1455                 0                 0                 1                 0   
1456                 0                 1                 0                 0   
1457                 0                 0                 1                 0   
1458                 0                 1                 0                 0   
1459                 1                 0                 0                 0   

      BsmtFinType2_GLQ  BsmtFinType2_LwQ  BsmtFinType2_Rec  BsmtFinType2_Unf  \
0                    0                 0                 0                 1   
1                    0                 0                 0                 1   
2                    0                 0                 0                 1   
3                    0                 0                 0                 1   
4                    0                 0                 0                 1   
...                ...               ...               ...               ...   
1455                 0                 0                 0                 1   
1456                 0                 0                 1                 0   
1457                 0                 0                 0                 1   
1458                 0                 0                 1                 0   
1459                 0                 1                 0                 0   

      BsmtFinType2_BLQ  BsmtFinType2_GLQ  BsmtFinType2_LwQ  BsmtFinType2_Rec  \
0                    0                 0                 0                 0   
1                    0                 0                 0   

In [321]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(exp_data,exp_labels,test_size=0.2)
print(f"x_train: {len(X_train)}\nx_test: {len(X_test)}")

x_train: 1136
x_test: 285


In [326]:
from sklearn.tree import DecisionTreeClassifier


dt = DecisionTreeClassifier().fit(X_train,y_train)
dt

ValueError: could not convert string to float: 'No'